File: scripts/check_nav_v2_task_edge_runtime_integration.py
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FILES = {
    'config': ROOT / 'config/nav-v2-task-edge-runtime-integration-v1.json',
    'manifest': ROOT / 'config/nav-v2-preview-deployment-bundle-manifest-v1.json',
    'decision': ROOT / 'config/nav-v2-deployment-decision-package-v1.json',
    'runtime': ROOT / 'supabase/functions/nav-v2-deal-api/task-action-edge-runtime-v2.js',
    'identity': ROOT / 'supabase/functions/nav-v2-deal-api/task-action-edge-identity-v2.js',
    'validator': ROOT / 'supabase/functions/nav-v2-deal-api/task-action-contract-v2.js',
    'index': ROOT / 'supabase/functions/nav-v2-deal-api/index.ts',
    'actor_sql': ROOT / 'supabase/prototypes/nav_v2_bounded_task_actor_aware_mutations.sql',
    'matrix': ROOT / 'scripts/check-nav-v2-task-edge-runtime-integration.mjs',
    'parity': ROOT / 'scripts/check-nav-v2-task-edge-sql-parity.mjs',
    'doc': ROOT / 'docs/NAV_V2_TASK_EDGE_RUNTIME_INTEGRATION_V1_2026-07-21.md',
    'workflow': ROOT / '.github/workflows/nav-v2-task-edge-runtime-integration-v1.yml',
    'guard': ROOT / 'assets/js/nav-v2/task-action-guard-v2.js',
}
MIGRATIONS = ROOT / 'supabase/migrations'


def need(text: str, markers: tuple[str, ...], label: str, errors: list[str]) -> None:
    for marker in markers:
        if marker not in text:
            errors.append(f'{label}: missing {marker!r}')


def main() -> int:
    errors: list[str] = []
    for label, path in FILES.items():
        if not path.exists():
            errors.append(f'missing {label}: {path.relative_to(ROOT)}')
    if errors:
        print('\n'.join(errors))
        return 1

    config = json.loads(FILES['config'].read_text(encoding='utf-8'))
    manifest = json.loads(FILES['manifest'].read_text(encoding='utf-8'))
    decision = json.loads(FILES['decision'].read_text(encoding='utf-8'))
    text = {label: path.read_text(encoding='utf-8') for label, path in FILES.items() if label not in ('config', 'manifest', 'decision')}

    if config.get('contract_version') != 1 or config.get('status') != 'repository_only_runtime_integration_disabled':
        errors.append('runtime integration contract metadata drifted')
    for key in ('production_applied', 'edge_deployed', 'feature_flag_default', 'frontend_transport_enabled', 'actor_aware_sql_deployed', 'service_role_key_browser_exposure_allowed'):
        if config.get(key) is not False:
            errors.append(f'{key} must remain false')
    if config.get('runtime_source_integrated') is not True or config.get('verify_jwt_required') is not True:
        errors.append('runtime source integration or JWT requirement missing')

    expected_actions = [
        'bounded_task_start', 'bounded_task_complete', 'bounded_task_active_outcome',
        'bounded_task_terminal_proposal', 'bounded_task_terminal_decision',
    ]
    if config.get('bounded_actions') != expected_actions:
        errors.append('bounded action inventory drifted')
    if config.get('identity_chain') != [
        'authorization_bearer_user_jwt', 'auth_v1_user_lookup', 'verified_auth_user_id',
        'active_nav_user_profile_lookup', 'contract_v2_task_context_lookup',
        'role_and_assignment_preflight', 'actor_aware_rpc_with_p_actor_id',
    ]:
        errors.append('identity chain drifted')

    role_policy = config.get('role_policy') or {}
    if role_policy.get('supervisors') != ['owner', 'admin', 'manager']:
        errors.append('supervisor role policy drifted')
    if role_policy.get('specialists') != ['spn', 'lawyer', 'broker']:
        errors.append('specialist role policy drifted')
    if role_policy.get('viewer_mutations_allowed') is not False:
        errors.append('viewer mutations must remain forbidden')
    broker = role_policy.get('broker') or {}
    if broker.get('task_type') != 'broker_task' or broker.get('allowed_sources') != ['intake_v1:mortgage', 'intake_v1:military_mortgage']:
        errors.append('broker mortgage-only policy drifted')

    if len(config.get('required_positive_scenarios') or []) != 4:
        errors.append('positive runtime scenario inventory drifted')
    if len(config.get('required_negative_scenarios') or []) < 11:
        errors.append('negative runtime scenario inventory is incomplete')
    for stop in ('feature_flag_disabled', 'actor_aware_sql_not_deployed', 'edge_not_deployed', 'frontend_transport_disabled', 'authenticated_e2e_not_proven', 'deployment_approval_missing'):
        if stop not in (config.get('active_stops') or []):
            errors.append(f'active stop missing: {stop}')
    for forbidden in ('service_role_key_in_frontend', 'service_role_key_in_response', 'service_role_key_in_log', 'client_supplied_actor_id', 'client_supplied_role', 'broker_non_mortgage_task', 'edge_deploy_from_repository_only_pr', 'enable_frontend_transport'):
        if forbidden not in (config.get('forbidden') or []):
            errors.append(f'forbidden runtime action missing: {forbidden}')

    need(text['runtime'], (
        "from './task-action-edge-identity-v2.js'",
        'routeBoundedTaskEdgeActionV2',
        "const SUPERVISOR_ROLES = new Set(['owner', 'admin', 'manager']);",
        "const SPECIALIST_ROLES = new Set(['spn', 'lawyer', 'broker']);",
        "const MORTGAGE_TASK_SOURCES = new Set(['intake_v1:mortgage', 'intake_v1:military_mortgage']);",
        "failed('feature_disabled'",
        "stage: 'runtime_rpc_executed'",
        "stage: 'broker_scope'",
        'task.task_contract_version !== 2',
        "actor_argument: 'p_actor_id'",
        'feature_flag_default: false',
        'edge_deployed: false',
        'frontend_transport_enabled: false',
    ), 'runtime adapter', errors)
    for forbidden in ('Deno.', 'fetch(', 'SUPABASE_SERVICE_ROLE_KEY', 'console.log', 'console.error'):
        if forbidden in text['runtime']:
            errors.append(f'runtime adapter contains forbidden transport/secret/log marker: {forbidden}')

    need(text['index'], (
        'import { routeBoundedTaskEdgeActionV2 } from "./task-action-edge-runtime-v2.js";',
        'const BOUNDED_TASK_EDGE_IDENTITY_ENABLED = false;',
        'boundedActions.has(action as BoundedNavV2Action)',
        'profile_loader: loadActiveNavProfile',
        'task_loader: loadBoundedTaskContext',
        'rpc_client: { rpc: callServiceRpc }',
        'SUPABASE_SERVICE_ROLE_KEY',
        'select=id,role,is_active&limit=1',
        'select=id,assigned_to,assigned_role,task_type,source,task_contract_version&limit=1',
    ), 'Edge index', errors)
    if text['index'].count('const BOUNDED_TASK_EDGE_IDENTITY_ENABLED = false;') != 1:
        errors.append('Edge feature flag count differs from one')
    if 'BOUNDED_TASK_EDGE_IDENTITY_ENABLED = true' in text['index']:
        errors.append('Edge feature flag was enabled')
    if re.search(r'jsonResponse\([^\n]*SUPABASE_SERVICE_ROLE_KEY', text['index']):
        errors.append('service role env appears in response path')
    if re.search(r'console\.(?:log|error|warn)\([^\n]*(?:serviceKey|SUPABASE_SERVICE_ROLE_KEY)', text['index']):
        errors.append('service role key appears in log path')

    need(text['identity'], (
        'FORBIDDEN_CLIENT_ACTOR_FIELDS',
        'const rpcArgs = { ...baseArgs, p_actor_id: actorId };',
        'actor_aware_sql_deployed: false',
    ), 'identity handler', errors)
    need(text['validator'], ('validateTaskEdgeAction', "action.startsWith('bounded_')"), 'action validator', errors)

    actor_signatures = re.findall(r'create\s+or\s+replace\s+function\s+public\.(nav_v2_[a-z0-9_]+)\s*\((.*?)\)\s*returns', text['actor_sql'], flags=re.I | re.S)
    actor_task_actions = []
    for name, arguments in actor_signatures:
        normalized = re.sub(r'\s+', ' ', arguments.lower())
        if 'p_actor_id uuid' in normalized and name != 'nav_v2_create_bounded_tasks':
            actor_task_actions.append(name.lower())
    if sorted(set(actor_task_actions)) != sorted([
        'nav_v2_start_bounded_task', 'nav_v2_complete_bounded_task',
        'nav_v2_set_bounded_task_active_outcome', 'nav_v2_propose_bounded_task_terminal_outcome',
        'nav_v2_decide_bounded_task_terminal_outcome',
    ]):
        errors.append('actor-aware SQL action signature inventory drifted')

    need(text['matrix'], (
        'spn_assigned_task', 'lawyer_assigned_task', 'broker_mortgage_task',
        'manager_supervisor_task', 'inactive_profile', 'viewer_mutation',
        'role_mismatch', 'assigned_other', 'broker_non_mortgage',
        'contract_v1', 'cross_actor_rpc_rejection',
        "calls.rpcArgs.args.p_actor_id, actor",
    ), 'runtime semantic matrix', errors)
    need(text['parity'], ('Edge-to-SQL parity passed', 'expected six actor-aware SQL overload definitions'), 'SQL parity checker', errors)
    need(text['guard'], ('const BOUNDED_TRANSPORT_ENABLED = false;',), 'frontend guard', errors)
    if 'SUPABASE_SERVICE_ROLE_KEY' in text['guard']:
        errors.append('service role env leaked into frontend guard')

    edge_layer = (manifest.get('layers') or [])[4]
    if manifest.get('edge_runtime_integrated') is not True or manifest.get('edge_runtime_enabled') is not False or manifest.get('edge_deployed') is not False:
        errors.append('preview manifest Edge state drifted')
    if edge_layer.get('mode') != 'source_integrated_feature_disabled' or edge_layer.get('edge_deploy_ready') is not False:
        errors.append('preview manifest Edge layer drifted')
    if decision.get('edge_runtime_source_integrated') is not True or decision.get('edge_runtime_enabled') is not False or decision.get('edge_deployed') is not False:
        errors.append('deployment decision Edge state drifted')
    if decision.get('deployment_bundle_ready') is not False:
        errors.append('deployment decision claims bundle readiness')

    for path in (ROOT / 'assets').rglob('*'):
        if path.is_file() and 'SUPABASE_SERVICE_ROLE_KEY' in path.read_text(encoding='utf-8', errors='ignore'):
            errors.append(f'service role env leaked into frontend asset: {path.relative_to(ROOT)}')
    leaked = [path.name for path in MIGRATIONS.glob('*edge*runtime*integration*')]
    if leaked:
        errors.append(f'Edge runtime integration leaked into migrations: {leaked}')

    need(text['doc'], (
        'repository-only Edge identity runtime integration',
        'Feature flag', 'Identity chain', 'Role policy', 'Broker scope',
        'Semantic matrix', 'Deployment gate', 'Rollback',
    ), 'runtime integration doc', errors)
    need(text['workflow'], (
        'check_nav_v2_task_edge_runtime_integration.py',
        'check-nav-v2-task-edge-runtime-integration.mjs',
        'check-nav-v2-task-edge-sql-parity.mjs',
        'check-nav-v2-task-edge-identity.mjs',
        'node --check supabase/functions/nav-v2-deal-api/task-action-edge-runtime-v2.js',
        'BOUNDED_TASK_EDGE_IDENTITY_ENABLED = false',
    ), 'runtime integration workflow', errors)

    if errors:
        print('Navigator v2 task Edge runtime integration errors:')
        for error in errors:
            print(f'- {error}')
        return 1

    print('Navigator v2 task Edge runtime integration source contract passed: verified actor route is integrated behind a disabled feature flag while SQL, deployment and frontend transport remain disabled')
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/check-nav-v2-task-edge-runtime-integration.mjs
import assert from 'node:assert/strict';
import { routeBoundedTaskEdgeActionV2 } from '../supabase/functions/nav-v2-deal-api/task-action-edge-runtime-v2.js';

const actor = '70000000-0000-4000-8000-000000000001';
const other = '70000000-0000-4000-8000-000000000002';
const taskId = '71000000-0000-4000-8000-000000000001';
const requestId = '72000000-0000-4000-8000-000000000001';

const requestBody = (payload = {}, action = 'bounded_task_start') => ({
  action,
  payload: {
    task_id: taskId,
    client_request_id: requestId,
    task_contract_version: 2,
    ...payload
  }
});

function profile(role = 'spn', overrides = {}) {
  return { id: actor, role, is_active: true, ...overrides };
}

function task(assignedRole = 'spn', overrides = {}) {
  return {
    id: taskId,
    assigned_to: actor,
    assigned_role: assignedRole,
    task_type: assignedRole === 'broker' ? 'broker_task' : 'operational_task',
    source: assignedRole === 'broker' ? 'intake_v1:mortgage' : 'intake_v1:test',
    task_contract_version: 2,
    ...overrides
  };
}

async function run({ enabled = true, body = requestBody(), user = { id: actor }, profileValue = profile(), taskValue = task(), rpcThrows = false } = {}) {
  const calls = { profile: 0, task: 0, rpc: 0, rpcArgs: null };
  const result = await routeBoundedTaskEdgeActionV2({
    enabled,
    request_body: body,
    verified_user: user,
    profile_loader: async (id) => {
      calls.profile += 1;
      assert.equal(id, actor);
      return profileValue;
    },
    task_loader: async (id) => {
      calls.task += 1;
      assert.equal(id, taskId);
      return taskValue;
    },
    rpc_client: {
      async rpc(name, args) {
        calls.rpc += 1;
        calls.rpcArgs = { name, args };
        if (rpcThrows) throw new Error('synthetic cross-actor rejection');
        return { ok: true, synthetic: true };
      }
    }
  });
  return { result, calls };
}

const disabled = await run({ enabled: false });
assert.equal(disabled.result.ok, false);
assert.equal(disabled.result.stage, 'feature_disabled');
assert.deepEqual(disabled.calls, { profile: 0, task: 0, rpc: 0, rpcArgs: null });

const accepted = [
  { id: 'spn_assigned_task', profileValue: profile('spn'), taskValue: task('spn') },
  { id: 'lawyer_assigned_task', profileValue: profile('lawyer'), taskValue: task('lawyer', { task_type: 'legal_blocker' }) },
  { id: 'broker_mortgage_task', profileValue: profile('broker'), taskValue: task('broker') },
  { id: 'manager_supervisor_task', profileValue: profile('manager'), taskValue: task('lawyer', { assigned_to: other, task_type: 'legal_blocker' }) }
];

for (const scenario of accepted) {
  const { result, calls } = await run(scenario);
  assert.equal(result.ok, true, `${scenario.id}: ${result.errors?.join(' | ')}`);
  assert.equal(result.stage, 'runtime_rpc_executed');
  assert.equal(result.runtime_integrated, true);
  assert.equal(result.route_enabled, true);
  assert.equal(result.edge_deployed, false);
  assert.equal(result.frontend_transport_enabled, false);
  assert.equal(result.verified_actor_id, actor);
  assert.equal(result.actor_role, scenario.profileValue.role);
  assert.deepEqual(calls.profile, 1);
  assert.deepEqual(calls.task, 1);
  assert.deepEqual(calls.rpc, 1);
  assert.equal(calls.rpcArgs.name, 'nav_v2_start_bounded_task');
  assert.equal(calls.rpcArgs.args.p_actor_id, actor);
  assert.equal(calls.rpcArgs.args.p_task_id, taskId);
  assert.equal(calls.rpcArgs.args.p_client_request_id, requestId);
}

const rejected = [
  { id: 'invalid_user', options: { user: { id: 'bad' } }, stage: 'verified_identity', calls: [0, 0, 0] },
  { id: 'forbidden_actor_field', options: { body: requestBody({ actor_id: actor }) }, stage: 'actor_trust_boundary', calls: [0, 0, 0] },
  { id: 'inactive_profile', options: { profileValue: profile('spn', { is_active: false }) }, stage: 'active_profile', calls: [1, 0, 0] },
  { id: 'unsupported_role', options: { profileValue: profile('unknown') }, stage: 'role_profile', calls: [1, 0, 0] },
  { id: 'viewer_mutation', options: { profileValue: profile('viewer'), taskValue: task('viewer') }, stage: 'role_policy', calls: [1, 1, 0] },
  { id: 'role_mismatch', options: { profileValue: profile('spn'), taskValue: task('lawyer') }, stage: 'role_mismatch', calls: [1, 1, 0] },
  { id: 'assigned_other', options: { profileValue: profile('lawyer'), taskValue: task('lawyer', { assigned_to: other }) }, stage: 'actor_assignment', calls: [1, 1, 0] },
  { id: 'broker_wrong_type', options: { profileValue: profile('broker'), taskValue: task('broker', { task_type: 'operational_task' }) }, stage: 'broker_scope', calls: [1, 1, 0] },
  { id: 'broker_non_mortgage', options: { profileValue: profile('broker'), taskValue: task('broker', { source: 'intake_v1:matcap' }) }, stage: 'broker_scope', calls: [1, 1, 0] },
  { id: 'contract_v1', options: { taskValue: task('spn', { task_contract_version: 1 }) }, stage: 'task_context', calls: [1, 1, 0] },
  { id: 'cross_actor_rpc_rejection', options: { rpcThrows: true }, stage: 'rpc_execution', calls: [1, 1, 1] }
];

for (const scenario of rejected) {
  const { result, calls } = await run(scenario.options);
  assert.equal(result.ok, false, `${scenario.id}: expected rejection`);
  assert.equal(result.stage, scenario.stage, `${scenario.id}: unexpected stage`);
  assert.deepEqual([calls.profile, calls.task, calls.rpc], scenario.calls, `${scenario.id}: unexpected call counts`);
  assert.equal(result.edge_deployed, false);
  assert.equal(result.frontend_transport_enabled, false);
}

console.log(`Navigator v2 Edge runtime integration matrix passed: ${accepted.length} accepted, ${rejected.length + 1} rejected, feature flag disabled by default`);
